### PR TITLE
[spreaker] Add new extractor

### PR DESCRIPTION
**Pull request by Tatsh**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Extractor for Spreaker.com podcast site. Fixes issue #13480

### CODE REVIEW

File: youtube_dl/extractor/spreaker.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+from ..compat import compat_str
+from ..utils import int_or_none, ExtractorError
+
+
+class SpreakerIE(InfoExtractor):
+    IE_NAME = 'spreaker'
+    _VALID_URL = r"""(?x)^
+        https?://
+        (?:www.|api.)?
+        spreaker.com/
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Dots must be escaped.

File: youtube_dl/extractor/spreaker.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        (?:
+            show/[a-z0-9_-]+|
+            user/[a-z0-9_-]+/[a-z0-9_-]|
+            episode/(?P<id>[0-9]+)
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Each should be a separate extractor.

File: youtube_dl/extractor/spreaker.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            'url': 'https://www.spreaker.com/show/success-with-music',
+            'info_dict': {
+                'title': 'Success With Music',
+                'id': 2317431,
+            },
+            'playlist_mincount': 14,
+        },
+        {
+            'url': ('https://www.spreaker.com/user/9780658/swm-ep15-how-to-'
+                    'market-your-music-part-2'),
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Don't carry URLs.

File: youtube_dl/extractor/spreaker.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                'title': 'Marketing Your Music - Part 2',
+                'upload_date': '20170809',
+                'uploader': 'SWM',
+                'uploader_id': 9780658,
+            },
+        }
+    ]
+
+    def _spreaker_episode_data_to_info(self, data):
+        upload_date = data['published_at'][0:10].replace('-', '')
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Should not break if `published_at` is missing.

File: youtube_dl/extractor/spreaker.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        author = data.get('author')
+        if not author:
+            author = {}
+        stats = data.get('stats')
+        view_count = like_count = comment_count = 0
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Lack of information is denoted by `None`.

File: youtube_dl/extractor/spreaker.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        show = data.get('show')
+        if not show:
+            show = {}
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: `show = data.get('show') or {}`

File: youtube_dl/extractor/spreaker.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        else:
+            show_image = show.get('image')
+        if not show_image:
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Reference before assignment for `show_image` if `not show`.

**Comment by Tatsh** [comment]
@dstftw This is ready to go on my end.

File: youtube_dl/extractor/spreaker.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        view_count = like_count = comment_count = None
+        show = data.get('show', {})
+        show_image = show.get('image', {})
+
+        if stats:
+            view_count = (stats.get('plays', 0) +
+                          stats.get('plays_streaming', 0) +
+                          stats.get('plays_download', 0))
+            like_count = stats.get('likes', 0)
+            comment_count = stats.get('messages', 0)
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Lack of information is denoted by `None`.

File: youtube_dl/extractor/spreaker.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+        urls = [x['api_url'] for x in items]
+        ret = []
+        for index, url in enumerate(urls):
+            data = self._download_json(url, None)['response']['episode']
+            dict_ = SpreakerIE._spreaker_episode_data_to_info(data)
+            dict_.update({
+                'playlist_id': compat_str(data['show_id']),
+                'playlist_title': data['show']['title'],
+                'playlist_index': index,
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: No such fields.

File: youtube_dl/extractor/spreaker.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+                break
+            items += new_items
+            next_url = playlist_.get('next_url')
+            page_no += 1
+
+        urls = [x['api_url'] for x in items]
+        ret = []
+        for index, url in enumerate(urls):
+            data = self._download_json(url, None)['response']['episode']
+            dict_ = SpreakerIE._spreaker_episode_data_to_info(data)
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: You must delegate with `url_result` instead.

File: youtube_dl/extractor/spreaker.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+    def _real_extract(self, url):
+        html = self._download_webpage(url, None)
+        episode_id = self._html_search_regex(
+            r'data-episode_id="(?P<id>[0-9]+)"', html, 'id')
+        if not re.match(r'^[0-9]+$', episode_id):
+            raise ExtractorError('Could not find episode ID')
+        data_url = 'https://api.spreaker.com/episode/%s' % (episode_id)
+
+        return self.url_result(data_url)
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Specify `ie_key` and `video_id`.

File: youtube_dl/extractor/spreaker.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            'ext': 'mp3',
+            'title': 'Marketing Your Music - Part 2',
+            'upload_date': '20170809',
+            'uploader': 'SWM',
+            'uploader_id': 9780658,
+        },
+    }
+
+    @staticmethod
+    def _spreaker_episode_data_to_info(data):
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: This is supposed to be in a class that does extraction, not in a class that delegates.

File: youtube_dl/extractor/spreaker.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+        url = 'https://api.spreaker.com/episode/%s' % (episode_id,)
+        data = self._download_json(url, episode_id)['response']['episode']
+        if not data['download_enabled']:
+            raise ExtractorError('Not supported yet')
+
+        return SpreakerIE._spreaker_episode_data_to_info(data)
+
+
+class SpreakerIE(InfoExtractor):
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Rename to something else.

File: youtube_dl/extractor/spreaker.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+            ret.append(dict_)
+
+        return self.playlist_result(ret,
+                                    data['show_id'],
+                                    data['show']['title'])
+
+
+class SpreakerAPIEpisodeIE(InfoExtractor):
+    IE_NAME = 'spreaker:api'
+    _VALID_URL = r'^https?://(?:api\.)?spreaker\.com/(?:download/)?episode/(?P<id>[0-9]+)(?:/[^\.]+\.mp3$)?'
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Remove ^.

File: youtube_dl/extractor/spreaker.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+                'playlist_index': index,
+            })
+            ret.append(dict_)
+
+        return self.playlist_result(ret,
+                                    data['show_id'],
+                                    data['show']['title'])
+
+
+class SpreakerAPIEpisodeIE(InfoExtractor):
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: Rename to `SpreakerIE`.

File: youtube_dl/extractor/spreaker.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+            'language': show.get('language'),
+            'thumbnails': [
+                {
+                    'id': show_image.get('image_id'),
+                    'url': show_image.get('big_url'),
+                    'width': int_or_none(show_image.get('width')),
+                    'height': int_or_none(show_image.get('height')),
+                },
+                {
+                    'url': show_image.get('large_url'),
```
Comment (dstftw) [inline youtube_dl/extractor/spreaker.py]: `url` should not be `None`.

**Comment by peterk** [comment]
@Tatsh I am also looking for this. Do you plan on finishing it? Was thinking about starting a new PR.

**Comment by Tatsh** [comment]
@peterk I pushed new code up with the requested changes from @dstftw. This should be good to go now.

I suggest maintaining a local branch of your own, merging master and merging the pull requests you like. This way you can use the newest features and stay up-to-date with master. Just don't use this branch for pull requests (always start with master).

**Comment by rudolphos** [comment]
Is there an update on this?

**Comment by Tatsh** [comment]
@dstftw Please review again since the last commit.

**Comment by Tatsh** [comment]
Bump

**Comment by rudolphos** [comment]
bump, still doesn't work

https://github.com/ytdl-org/youtube-dl/issues/13480#issuecomment-529050961

**Comment by Tatsh** [comment]
@rudolphos This version works fine.

**Comment by rudolphos** [comment]
I'm using the latest version and it doesn't work 

```
Input URL: https://www.spreaker.com/user/hjmadigan/51-lost-in-a-good-game --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-ci', '--write-thumbnail', '--no-check-certificate', '--sub-format', 'ass/srt/best', '--write-auto-sub', '--convert-subs', 'srt', '--write-description', '-f', 'bestvideo+bestaudio/best', '--download-archive', 'downloaded.txt', '--output', '/Videos/%(uploader)s - %(title)s (%(id)s).%(ext)s', 'https://www.spreaker.com/user/hjmadigan/51-lost-in-a-good-game', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2019.09.12.1
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.18362
[debug] exe versions: ffmpeg N-93825-gc967128952, ffprobe N-93825-gc967128952, rtmpdump 2.4
[debug] Proxy map: {}
[generic] 51-lost-in-a-good-game: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 51-lost-in-a-good-game: Downloading webpage
[generic] 51-lost-in-a-good-game: Extracting information
[generic] 8913843fbb3c94807f9351c73d45ca5e: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 8913843fbb3c94807f9351c73d45ca5e: Downloading webpage
[generic] 8913843fbb3c94807f9351c73d45ca5e: Extracting information
ERROR: Unsupported URL: https://widget.spreaker.com/player?episode_id=19032297&autoplay=false&playlist=show&cover_image_url=https://d3wo5wojvuv7l.cloudfront.net/images.spreaker.com/original/8913843fbb3c94807f9351c73d45ca5e.jpg
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\extractor\common.py", line 530, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\extractor\generic.py", line 3355, in _real_extract
youtube_dl.utils.UnsupportedError: Unsupported URL: https://widget.spreaker.com/player?episode_id=19032297&autoplay=false&playlist=show&cover_image_url=https://d3wo5wojvuv7l.cloudfront.net/images.spreaker.com/original/8913843fbb3c94807f9351c73d45ca5e.jpg
```

```
Input URL: https://www.spreaker.com/show/psychology-of-video-games-podcast_1 --verbose
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-ci', '--write-thumbnail', '--no-check-certificate', '--sub-format', 'ass/srt/best', '--write-auto-sub', '--convert-subs', 'srt', '--write-description', '-f', 'bestvideo+bestaudio/best', '--download-archive', 'downloaded.txt', '--output', '/Videos/%(uploader)s - %(title)s (%(id)s).%(ext)s', 'https://www.spreaker.com/show/psychology-of-video-games-podcast_1', '--verbose']
[debug] Encodings: locale cp1252, fs mbcs, out cp437, pref cp1252
[debug] youtube-dl version 2019.09.12.1
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.18362
[debug] exe versions: ffmpeg N-93825-gc967128952, ffprobe N-93825-gc967128952, rtmpdump 2.4
[debug] Proxy map: {}
[generic] psychology-of-video-games-podcast_1: Requesting header
WARNING: Falling back on generic information extractor.
[generic] psychology-of-video-games-podcast_1: Downloading webpage
[generic] psychology-of-video-games-podcast_1: Extracting information
[generic] 6773d4af57b93eac576d2c1ab827bf40: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 6773d4af57b93eac576d2c1ab827bf40: Downloading webpage
[generic] 6773d4af57b93eac576d2c1ab827bf40: Extracting information
ERROR: Unsupported URL: https://widget.spreaker.com/player?show_id=3329893&autoplay=false&cover_image_url=https://d1bm3dmew779uf.cloudfront.net/cover/6773d4af57b93eac576d2c1ab827bf40.jpg
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\extractor\common.py", line 530, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmphopugs4t\build\youtube_dl\extractor\generic.py", line 3355, in _real_extract
youtube_dl.utils.UnsupportedError: Unsupported URL: https://widget.spreaker.com/player?show_id=3329893&autoplay=false&cover_image_url=https://d1bm3dmew779uf.cloudfront.net/cover/6773d4af57b93eac576d2c1ab827bf40.jpg
```

**Comment by Tatsh** [comment]
@dstftw If you want me to make changes, I'll need to re-make this PR. I asked GitHub and they cannot fix branch references.

**Comment by dstftw** [comment]
Picked some code in 686e898. Thanks.

**Pull request closed without merging**